File: frontend/src/engine/chemistry.ts
/**
 * Element contents of minerals from formulas and standard atomic weights (port of engine/chemistry.py).
 * Formulas allow decimal subscripts and nested parentheses; oxide grades such as P2O5 and MgO are
 * reported from their element through the oxide's own formula weight.
 */
import { atomicWeights, mineralTable, oxideTable } from './constants';

const TOKEN = /([A-Z][a-z]?|\(|\)|\d+(?:\.\d+)?)/g;
const isDigit = (token: string) => token[0] >= '0' && token[0] <= '9';

export function parseFormula(formula: string): Record<string, number> {
  const clean = formula.replace(/ /g, '');
  const tokens = clean.match(TOKEN) ?? [];
  if (tokens.join('') !== clean) throw new Error(`unparseable formula: ${formula}`);
  const stack: Array<Record<string, number>> = [{}];
  let i = 0;
  while (i < tokens.length) {
    const token = tokens[i];
    if (token === '(') { stack.push({}); i += 1; continue; }
    if (token === ')') {
      const group = stack.pop() as Record<string, number>;
      let count = 1.0;
      if (i + 1 < tokens.length && isDigit(tokens[i + 1])) { count = Number(tokens[i + 1]); i += 1; }
      const top = stack[stack.length - 1];
      for (const [element, n] of Object.entries(group)) top[element] = (top[element] ?? 0.0) + n * count;
      i += 1;
      continue;
    }
    if (isDigit(token)) throw new Error(`misplaced count in formula: ${formula}`);
    let count = 1.0;
    if (i + 1 < tokens.length && isDigit(tokens[i + 1])) { count = Number(tokens[i + 1]); i += 1; }
    const top = stack[stack.length - 1];
    top[token] = (top[token] ?? 0.0) + count;
    i += 1;
  }
  if (stack.length !== 1) throw new Error(`unbalanced parentheses: ${formula}`);
  return stack[0];
}

export function formulaWeight(formula: string): number {
  let total = 0.0;
  for (const [element, n] of Object.entries(parseFormula(formula))) total += atomicWeights[element] * n;
  return total;
}

export function massFractions(formula: string): Record<string, number> {
  const counts = parseFormula(formula);
  let total = 0.0;
  for (const [element, n] of Object.entries(counts)) total += atomicWeights[element] * n;
  const out: Record<string, number> = {};
  for (const [element, n] of Object.entries(counts)) out[element] = atomicWeights[element] * n / total;
  return out;
}

export function oxideFactor(oxide: string): number {
  const entry = oxideTable[oxide];
  const counts = parseFormula(entry.formula);
  return formulaWeight(entry.formula) / (atomicWeights[entry.element] * counts[entry.element]);
}

const compositionCache = new Map<string, Record<string, number>>();

export function mineralComposition(mineralId: string): Record<string, number> {
  const cached = compositionCache.get(mineralId);
  if (cached) return cached;
  const entry = mineralTable[mineralId];
  let out: Record<string, number>;
  if (entry.formula) out = massFractions(entry.formula);
  else {
    const composition = entry.composition as Record<string, number>;
    let total = 0.0;
    for (const v of Object.values(composition)) total += v;
    out = {};
    for (const [element, value] of Object.entries(composition)) out[element] = value / total;
  }
  compositionCache.set(mineralId, out);
  return out;
}

export function speciesContent(composition: Record<string, number>, species: string): number {
  if (species in composition) return composition[species];
  if (species in oxideTable) return (composition[oxideTable[species].element] ?? 0.0) * oxideFactor(species);
  return 0.0;
}

File: data-pipeline/pipeline/engine/chemistry.py
"""Element contents of minerals from formulas and standard atomic weights.

A formula such as ``Ca5(PO4)3F`` or ``Fe4.5Ni4.5S8`` is parsed into element counts (decimal
subscripts and nested parentheses are allowed) and converted to mass fractions. Oxide grades such
as P2O5 and MgO are reported from their element through the oxide's own formula weight.
"""
from __future__ import annotations

import re
from functools import lru_cache

from .constants import atomic_weights, mineral_table, oxide_table

_TOKEN = re.compile(r"([A-Z][a-z]?|\(|\)|\d+(?:\.\d+)?)")


def parse_formula(formula: str) -> dict[str, float]:
    """Return element counts per formula unit."""
    tokens = _TOKEN.findall(formula.replace(" ", ""))
    if "".join(tokens) != formula.replace(" ", ""):
        raise ValueError(f"unparseable formula: {formula}")
    stack: list[dict[str, float]] = [{}]
    i = 0
    while i < len(tokens):
        token = tokens[i]
        if token == "(":
            stack.append({})
            i += 1
            continue
        if token == ")":
            group = stack.pop()
            count = 1.0
            if i + 1 < len(tokens) and tokens[i + 1][0].isdigit():
                count = float(tokens[i + 1])
                i += 1
            for element, n in group.items():
                stack[-1][element] = stack[-1].get(element, 0.0) + n * count
            i += 1
            continue
        if token[0].isdigit():
            raise ValueError(f"misplaced count in formula: {formula}")
        count = 1.0
        if i + 1 < len(tokens) and tokens[i + 1][0].isdigit():
            count = float(tokens[i + 1])
            i += 1
        stack[-1][token] = stack[-1].get(token, 0.0) + count
        i += 1
    if len(stack) != 1:
        raise ValueError(f"unbalanced parentheses: {formula}")
    return stack[0]


def formula_weight(formula: str) -> float:
    weights = atomic_weights()
    return sum(weights[element] * n for element, n in parse_formula(formula).items())


def mass_fractions(formula: str) -> dict[str, float]:
    weights = atomic_weights()
    counts = parse_formula(formula)
    total = sum(weights[element] * n for element, n in counts.items())
    return {element: weights[element] * n / total for element, n in counts.items()}


def oxide_factor(oxide: str) -> float:
    """Mass of oxide per unit mass of its element (for example P2O5 per P)."""
    entry = oxide_table()[oxide]
    element = entry["element"]
    counts = parse_formula(entry["formula"])
    return formula_weight(entry["formula"]) / (atomic_weights()[element] * counts[element])


@lru_cache(maxsize=None)
def mineral_composition(mineral_id: str) -> dict[str, float]:
    """Element mass fractions of a library mineral."""
    entry = mineral_table()[mineral_id]
    if "formula" in entry:
        return mass_fractions(entry["formula"])
    composition = dict(entry["composition"])
    total = sum(composition.values())
    return {element: value / total for element, value in composition.items()}


def species_content(composition: dict[str, float], species: str) -> float:
    """Mass fraction of an element or an oxide species (P2O5, MgO, SiO2) in a composition."""
    if species in composition:
        return composition[species]
    oxides = oxide_table()
    if species in oxides:
        element = oxides[species]["element"]
        return composition.get(element, 0.0) * oxide_factor(species)
    return 0.0
